platform/linux-generic/odp_ipsec.c

```diff
@@ -1055,7 +1069,16 @@ int odp_ipsec_in_enq(const odp_packet_t pkt_in[], int num_in,
 		result->status = status;
 		if (NULL != ipsec_sa) {
 			result->sa = ipsec_sa->ipsec_sa_hdl;
-			queue = ipsec_sa->queue;
+			if (ipsec_sa->in.cos && !status.error.all) {
+				odp_packet_hdr_t *pkt_hdr = odp_packet_hdr(pkt);
+				const uint8_t *base = odp_packet_data(pkt);
+
+				queue = cls_pkt_get_queue(pkt_hdr,
+							  ipsec_sa->in.cos,
+							  base);
+			} else {
```

Comment:
Pipelining isn't about sending the result to an alternate queue that's known ahead of time, but rather about sending the result to the `cls_classify_packet()` function so that classification can be run against the decrypted packet and PMRs can filter it into the correct CoS (and associated queue) that is to receive the packet.

This is likely to be a sufficiently large change in itself that it may be better to be its own PR, in which case this can be pulled from the rest of this PR so that the other changes can be merged first.

Reply:
@Bill-Fischofer-Linaro well, this is not what API says. Currently if `ODP_PIPELINE_CLS` is selected, IPsec packets go to predefined (per-SA) `odp_cos_t`.

Reply:
@Bill-Fischofer-Linaro quoting spec:
```
                         *  Successfully decapsulated packets are sent to
                         *  classification through this CoS. Other resulting
                         *  events are sent to 'dest_queue'. This field is
                         *  considered only when 'pipeline' is
                         *  ODP_IPSEC_PIPELINE_CLS. The CoS must not be shared
                         *  between any pktio interface default CoS. The maximum
                         *  number of different CoS supported is defined by
                         *  IPSEC capability max_cls_cos.

```

Reply:
Read that again. It specifies a CoS that begins classification, not a queue. Putting a packet on a queue does not trigger the classifier. The current linux-generic classifier implementation doesn't support this type of "alternate entry point", nor does it support decrypted packets that probably don't start with L2 headers, etc. That's why I suggest this should probably be its own PR because it will involve enhancements to the classifier as well as IPsec to achieve the desired result.

Reply:
OK, sorry, I was confused by the text. Dropped these patches for now.